xds/src/main/java/io/grpc/xds/XdsLoadReportStore.java

```diff
@@ -0,0 +1,166 @@
+      }
+    }
+    for (Map.Entry<String, AtomicLong> entry : dropCounters.entrySet()) {
+      statsBuilder.addDroppedRequests(DroppedRequests.newBuilder()
+          .setCategory(entry.getKey())
+          .setDroppedCount(entry.getValue().get()));
+    }
+    return statsBuilder.build();
+  }
+
+  /**
+   * Create a {@link ClientLoadCounter} for the provided locality or make it active if already in
+   * this {@link XdsLoadReportStore}. This method needs to be called at locality updates only for
+   * newly assigned localities in balancer discovery responses.
+   * This method should be called in the same synchronized context that
+   * {@link XdsLoadBalancer#helper#getSynchronizationContext} returns.
+   */
+  void addLocality(final Locality locality) {
+    ClientLoadCounter counter = localityLoadCounters.get(locality);
+    checkState(counter == null || !counter.isActive(),
+        "An active ClientLoadCounter for locality %s already exists", locality);
+    if (counter == null) {
+      localityLoadCounters.put(locality, new ClientLoadCounter());
+    } else {
+      counter.setActive(true);
+    }
+  }
+
+  /**
+   * Deactivate the {@link ClientLoadCounter} for the provided locality in by this
+   * {@link XdsLoadReportStore}. Inactive {@link ClientLoadCounter}s are for localities
+   * no longer exposed by the remote balancer. This method needs to be called at
+   * locality updates only for localities newly removed from balancer discovery responses.
+   * This method should be called in the same synchronized context that
+   * {@link XdsLoadBalancer#helper#getSynchronizationContext} returns.
+   */
+  void removeLocality(final Locality locality) {
+    ClientLoadCounter counter = localityLoadCounters.get(locality);
+    checkState(counter != null && counter.isActive(),
+        "No active ClientLoadCounter for locality %s exists", locality);
+    counter.setActive(false);
+  }
+
+  /**
+   * Intercepts a in-locality PickResult with load recording {@link ClientStreamTracer.Factory}.
+   */
+  PickResult interceptPickResult(PickResult pickResult, Locality locality) {
+    if (!pickResult.getStatus().isOk()) {
+      return pickResult;
+    }
+    XdsClientLoadRecorder.ClientLoadCounter counter = localityLoadCounters.get(locality);
+    if (counter == null) {
+      return pickResult;
+    }
+    ClientStreamTracer.Factory originFactory = pickResult.getStreamTracerFactory();
+    if (originFactory == null) {
+      originFactory = NOOP_CLIENT_STREAM_TRACER_FACTORY;
+    }
+    XdsClientLoadRecorder recorder = new XdsClientLoadRecorder(counter, originFactory);
+    return PickResult.withSubchannel(pickResult.getSubchannel(), recorder);
+  }
+
+  /**
+   * Record that a request has been dropped by drop overload policy with the provided category
+   * instructed by the remote balancer.
+   */
+  void recordDroppedRequest(String category) {
+    AtomicLong counter = dropCounters.putIfAbsent(category, new AtomicLong());
```

Comment:
For counting dropped requests in xDS, I am thinking that we could have a `DroppedPickResult` that extends `PickResult` with an extra field `reason` for the category of this drop belongs to. `PickResult` already has a method `isDrop` that tells if the result is a drop. Then we could gracefully handle the dropped requests counting in `interceptPickResult` method by first checking `pickResult.isDrop()`, if it is a drop, we cast it to `DroppedPickResult`, call `getReason()` on it to get the drop reason and update the corresponding counter in `dropCounters` map. With this approach, the xDS policy does not need to call `recordDroppedRequest` to count it. The only ugliness of this approach is the down casting, but in this case, I think it is totally fine.

Reply:
`PickResult` is final so it's not ready to be extended. `isDrop()` is on it because that signal is used by gRPC library to determine its retry eligibility. It's not the case for the drop category. Having drop-counting in `XdsLoadReportStore` instead of the xDS policy doesn't sound a good enough reason to make the API change on `PickResult`.